File: apps/docker-server/cmd/renewlet/subscription_public_visibility.go
package main

import (
	"encoding/json"
	"errors"
	"net/http"
	"strings"
	"unicode/utf8"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tools/security"
	"github.com/pocketbase/pocketbase/tools/types"
)

var errSubscriptionWriteConflict = errors.New("SUBSCRIPTION_WRITE_CONFLICT")
var errPublicVisibilityIndex = errors.New("SUBSCRIPTION_PUBLIC_INDEX_INCONSISTENT")

type subscriptionBulkPublicVisibilitySelection struct {
	IDs        []string `json:"ids,omitempty"`
	Categories []string `json:"categories,omitempty"`
}

type subscriptionBulkPublicVisibilityRequest struct {
	Selection    subscriptionBulkPublicVisibilitySelectionInput `json:"selection"`
	PublicHidden optionalJSONField[bool]                        `json:"publicHidden"`
	DryRun       optionalJSONField[bool]                        `json:"dryRun"`
}

type subscriptionBulkPublicVisibilitySelectionInput struct {
	IDs        optionalJSONField[[]string] `json:"ids"`
	Categories optionalJSONField[[]string] `json:"categories"`
}

type subscriptionBulkPublicVisibilityResponse struct {
	MatchedCount int      `json:"matchedCount"`
	ChangedCount int      `json:"changedCount"`
	SkippedCount int      `json:"skippedCount"`
	FailedIDs    []string `json:"failedIds"`
}

type publicVisibilityTarget struct {
	ID               string `db:"id" json:"id"`
	User             string `db:"user" json:"user"`
	PublicHidden     bool   `db:"publicHidden" json:"publicHidden"`
	Updated          string `db:"updated" json:"updated"`
	Status           string `db:"status" json:"status"`
	NextBillingDate  string `db:"nextBillingDate" json:"nextBillingDate"`
	BillingCycle     string `db:"billingCycle" json:"billingCycle"`
	OneTimeTermCount int    `db:"oneTimeTermCount" json:"oneTimeTermCount"`
}

const visibilityColumns = "s.id, s.user, s.publicHidden, s.updated, s.status, s.nextBillingDate, s.billingCycle, s.oneTimeTermCount"

func handleBulkPublicVisibility(app core.App, e *core.RequestEvent) error {
	locale := requestLocale(e.Request)
	body, err := decodeStrictJSON[subscriptionBulkPublicVisibilityRequest](e.Request, locale)
	if err != nil || !body.PublicHidden.Set || body.PublicHidden.Null || body.DryRun.Null {
		return e.BadRequestError(serverText(locale, "common.invalidRequestBody"), err)
	}
	// 与 shared 严格互斥联合一致：显式 null 也是已提供字段，不能当作省略而接受两种选择范围。
	if body.Selection.IDs.Set == body.Selection.Categories.Set || body.Selection.IDs.Null || body.Selection.Categories.Null {
		return e.BadRequestError(serverText(locale, "common.invalidRequestParameters"), nil)
	}
	selection := subscriptionBulkPublicVisibilitySelection{IDs: body.Selection.IDs.Value, Categories: body.Selection.Categories.Value}
	if (len(selection.IDs) == 0 && len(selection.Categories) == 0) || len(selection.IDs) > subscriptionCollectionLimit || len(selection.Categories) > 2 {
		return e.BadRequestError(serverText(locale, "common.invalidRequestParameters"), nil)
	}
	for _, id := range selection.IDs {
		if value := strings.TrimSpace(id); value == "" || utf8.RuneCountInString(value) > 80 {
			return e.BadRequestError(serverText(locale, "common.invalidRequestParameters"), nil)
		}
	}
	for _, category := range selection.Categories {
		if category != "expired" && category != "lifetime" {
			return e.BadRequestError(serverText(locale, "common.invalidRequestParameters"), nil)
		}
	}
	selection.IDs = uniqueNonEmptyStrings(trimVisibilityIDs(selection.IDs)...)
	rows, err := readPublicVisibilityTargets(app, e.Auth.Id, selection, subscriptionQueryToday(app, e.Auth, subscriptionListQuery{}))
	if err != nil {
		return publicVisibilityError(app, e, err)
	}
	if len(rows) > subscriptionCollectionLimit {
		return apiErrorJSON(e, http.StatusUnprocessableEntity, "SUBSCRIPTION_COLLECTION_LIMIT_EXCEEDED", serverText(locale, "common.invalidRequestParameters"), map[string]int{"limit": subscriptionCollectionLimit})
	}
	changed := make([]publicVisibilityTarget, 0, len(rows))
	matched := make(map[string]bool, len(rows))
	for _, row := range rows {
		matched[row.ID] = true
		if row.PublicHidden != body.PublicHidden.Value {
			changed = append(changed, row)
		}
	}
	failed := []string{}
	for _, id := range selection.IDs {
		if !matched[id] {
			failed = append(failed, id)
		}
	}
	if !body.DryRun.Value {
		if err := writePublicVisibility(app, e.Auth.Id, changed, body.PublicHidden.Value); err != nil {
			return publicVisibilityError(app, e, err)
		}
	}
	return apiSuccessJSON(e, http.StatusOK, subscriptionBulkPublicVisibilityResponse{len(rows), len(changed), len(rows) - len(changed), failed})
}

func trimVisibilityIDs(ids []string) []string {
	result := make([]string, len(ids))
	for i, id := range ids {
		result[i] = strings.TrimSpace(id)
	}
	return result
}

func readPublicVisibilityTargets(app core.App, userID string, selection subscriptionBulkPublicVisibilitySelection, today string) ([]publicVisibilityTarget, error) {
	rows := []publicVisibilityTarget{}
	params := dbx.Params{"user": userID, "today": today}
	query := "SELECT " + visibilityColumns + " FROM subscriptions AS s WHERE s.user = {:user}"
	if len(selection.IDs) > 0 {
		ids, err := json.Marshal(selection.IDs)
		if err != nil {
			return nil, err
		}
		params["ids"] = string(ids)
		// JSON 目标表必须驱动 ID 点查；允许优化器先遍历 owner 索引会把每个目标重复扫描成 O(n²)。
		query = "SELECT " + visibilityColumns + " FROM json_each({:ids}) AS selected CROSS JOIN subscriptions AS s ON s.id = selected.value WHERE s.user = {:user} ORDER BY CAST(selected.key AS INTEGER)"
	} else {
		clauses := []string{}
		for _, category := range uniqueNonEmptyStrings(selection.Categories...) {
			if category == "expired" {
				clauses = append(clauses, "(s.status = 'expired' OR (s.status IN ('active', 'trial') AND s.nextBillingDate < {:today}))")
			}
			if category == "lifetime" {
				clauses = append(clauses, "(s.billingCycle = 'one-time' AND s.oneTimeTermCount <= 0)")
			}
		}
		query += " AND (" + strings.Join(clauses, " OR ") + ") ORDER BY s.created DESC, s.id DESC LIMIT 5001"
	}
	err := app.DB().NewQuery(query).Bind(params).All(&rows)
	return rows, err
}

func writePublicVisibility(app core.App, userID string, rows []publicVisibilityTarget, publicHidden bool) error {
	if len(rows) == 0 {
		return nil
	}
	if len(rows) > subscriptionCollectionLimit {
		return errors.New("visibility mutation exceeds limit")
	}
	for _, row := range rows {
		if row.User != userID {
			return errSubscriptionWriteConflict
		}
	}
	payload, err := json.Marshal(rows)
	if err != nil {
		return err
	}
	params := dbx.Params{"rows": string(payload), "user": userID, "target": boolToSQLiteInt(publicHidden), "now": types.NowDateTime().String()}
	// 单条公开性 PATCH 与批量命令共用事务；只写事实公开列和投影，不重建标签、统计或通知调度。
	// PocketBase 时间必须使用 DateTime.String；RFC3339 直写会破坏原始快照和后续日期解析的一致性。
	return app.RunInTransaction(func(txApp core.App) error {
		var counts struct {
			Facts      int `db:"facts"`
			Projection int `db:"projection"`
		}
		err := txApp.DB().NewQuery(`SELECT COUNT(*) AS facts,
			COALESCE(SUM(CASE WHEN i.user_id = s.user AND i.public_hidden = s.publicHidden THEN 1 ELSE 0 END), 0) AS projection
			FROM json_each({:rows}) AS expected CROSS JOIN subscriptions AS s ON s.id = json_extract(expected.value, '$.id')
			LEFT JOIN subscription_list_index AS i ON i.subscription_id = s.id WHERE s.user = {:user}
			AND s.publicHidden IS json_extract(expected.value, '$.publicHidden') AND s.updated IS json_extract(expected.value, '$.updated')
			AND s.status IS json_extract(expected.value, '$.status') AND s.nextBillingDate IS json_extract(expected.value, '$.nextBillingDate')
			AND s.billingCycle IS json_extract(expected.value, '$.billingCycle') AND s.oneTimeTermCount IS json_extract(expected.value, '$.oneTimeTermCount')`).Bind(params).One(&counts)
		if err != nil {
			return err
		}
		if counts.Facts != len(rows) {
			return errSubscriptionWriteConflict
		}
		if counts.Projection != len(rows) {
			return errPublicVisibilityIndex
		}
		for _, sql := range []string{
			`UPDATE subscriptions SET publicHidden = {:target}, updated = {:now} WHERE user = {:user} AND id IN (SELECT json_extract(value, '$.id') FROM json_each({:rows}))`,
			`UPDATE subscription_list_index SET public_hidden = {:target}, updated_at = {:now} WHERE user_id = {:user} AND subscription_id IN (SELECT json_extract(value, '$.id') FROM json_each({:rows}))`,
		} {
			result, err := txApp.DB().NewQuery(sql).Bind(params).Execute()
			if err != nil {
				return err
			}
			count, err := result.RowsAffected()
			if err != nil {
				return err
			}
			if count != int64(len(rows)) {
				return errPublicVisibilityIndex
			}
		}
		var applied struct {
			Count int `db:"count"`
		}
		if err := txApp.DB().NewQuery(`SELECT COUNT(*) AS count FROM json_each({:rows}) AS selected
            CROSS JOIN subscriptions AS s ON s.id=json_extract(selected.value, '$.id')
            JOIN subscription_list_index AS i ON i.subscription_id=s.id
            WHERE s.user={:user} AND i.user_id=s.user AND s.publicHidden={:target} AND i.public_hidden=s.publicHidden
              AND s.updated={:now} AND i.updated_at=s.updated`).Bind(params).One(&applied); err != nil {
			return err
		}
		if applied.Count != len(rows) {
			return errPublicVisibilityIndex
		}
		return nil
	})
}

func visibilityTargetFromRecord(record *core.Record) publicVisibilityTarget {
	return publicVisibilityTarget{record.Id, record.GetString("user"), record.GetBool("publicHidden"), record.GetDateTime("updated").String(), record.GetString("status"), record.GetString("nextBillingDate"), record.GetString("billingCycle"), record.GetInt("oneTimeTermCount")}
}

func handleSinglePublicVisibilityUpdate(app core.App, e *core.RequestEvent, record *core.Record, target optionalJSONField[bool]) error {
	if target.Null {
		return e.BadRequestError(serverText(requestLocale(e.Request), "common.invalidRequestParameters"), nil)
	}
	if record.GetBool("publicHidden") != target.Value {
		if err := writePublicVisibility(app, e.Auth.Id, []publicVisibilityTarget{visibilityTargetFromRecord(record)}, target.Value); err != nil {
			return publicVisibilityError(app, e, err)
		}
		current, err := findOwnedSubscription(app, e)
		if err != nil {
			return publicVisibilityError(app, e, err)
		}
		record = current
	}
	return apiSuccessJSON(e, http.StatusOK, subscriptionResponse{Subscription: subscriptionAPIFromRecord(record)})
}

func publicVisibilityError(app core.App, e *core.RequestEvent, err error) error {
	locale := requestLocale(e.Request)
	if errors.Is(err, errSubscriptionWriteConflict) {
		return apiErrorJSON(e, http.StatusConflict, "SUBSCRIPTION_WRITE_CONFLICT", serverText(locale, "subscription.visibilityConflict"), nil)
	}
	requestID := security.RandomString(16)
	app.Logger().Error("subscription_public_visibility_failed", "requestId", requestID, "error", err)
	return apiErrorJSON(e, http.StatusInternalServerError, "SUBSCRIPTION_PUBLIC_VISIBILITY_FAILED", serverText(locale, "subscription.visibilityFailed"), map[string]string{"requestId": requestID})
}

File: apps/worker/src/subscription-public-visibility.ts
import { SUBSCRIPTION_INDEX_LIMIT, subscriptionBulkPublicVisibilityPayloadSchema, subscriptionBulkPublicVisibilityRequestSchema, type SubscriptionBulkPublicVisibilityRequest } from "@renewlet/shared/schemas/subscriptions";
import { requireAuth } from "./auth";
import { boolToInt, getSettings, nowIso } from "./db";
import { HttpError, readJson, requestLocale, successJson, type AppLocale } from "./http";
import { serverText } from "./server-i18n";
import { dateOnlyInZone } from "./subscription-renewal";
import type { Env, SubscriptionRow } from "./types";

const targetColumns = ["id", "user_id", "public_hidden", "updated_at", "status", "next_billing_date", "billing_cycle", "one_time_term_count"] as const;
type VisibilityTarget = Pick<SubscriptionRow, typeof targetColumns[number]>;
const snapshotColumns = targetColumns.filter((column) => column !== "user_id");

export async function bulkUpdatePublicVisibility(request: Request, env: Env): Promise<Response> {
  const locale = requestLocale(request);
  const auth = await requireAuth(request, env);
  const body = await readJson(request, subscriptionBulkPublicVisibilityRequestSchema, locale);
  try {
    const today = dateOnlyInZone(new Date(), (await getSettings(env, auth.user.id)).timezone);
    const rows = await listVisibilityTargets(env, auth.user.id, body.selection, today, locale);
    const matchedIds = new Set(rows.map((row) => row.id));
    const failedIds = "ids" in body.selection ? [...new Set(body.selection.ids)].filter((id) => !matchedIds.has(id)) : [];
    const changed = rows.filter((row) => row.public_hidden !== boolToInt(body.publicHidden));
    if (!body.dryRun) await writePublicVisibility(env, auth.user.id, changed, body.publicHidden, locale);
    return successJson(subscriptionBulkPublicVisibilityPayloadSchema.parse({
      matchedCount: rows.length, changedCount: changed.length, skippedCount: rows.length - changed.length, failedIds,
    }));
  } catch (error) {
    throw visibilityError(error, locale);
  }
}

async function listVisibilityTargets(env: Env, userId: string, selection: SubscriptionBulkPublicVisibilityRequest["selection"], today: string, locale: AppLocale): Promise<VisibilityTarget[]> {
  if ("ids" in selection) {
    const ids = [...new Set(selection.ids)];
    const byId = new Map<string, VisibilityTarget>();
    // D1 每条语句最多 100 个绑定参数；400 条读取批次用 JSON1 传两个参数，不能为每个 ID 展开占位符。
    // JSON 目标驱动 ID 点查，避免每个分块重新扫描 owner 下的全部订阅。
    for (let offset = 0; offset < ids.length; offset += 400) {
      const result = await env.DB.prepare(`SELECT ${targetColumns.map((column) => `s.${column}`).join(", ")}
        FROM json_each(?) AS selected CROSS JOIN subscriptions AS s ON s.id = selected.value WHERE s.user_id = ?`)
        .bind(JSON.stringify(ids.slice(offset, offset + 400)), userId).all<VisibilityTarget>();
      for (const row of result.results) byId.set(row.id, row);
    }
    return ids.flatMap((id) => { const row = byId.get(id); return row ? [row] : []; });
  }
  const predicates: string[] = [];
  const values: Array<string | number> = [userId];
  if (selection.categories.includes("expired")) {
    predicates.push("(status = 'expired' OR (status IN ('active', 'trial') AND next_billing_date < ?))");
    values.push(today);
  }
  if (selection.categories.includes("lifetime")) predicates.push("(billing_cycle = 'one-time' AND COALESCE(one_time_term_count, 0) <= 0)");
  const result = await env.DB.prepare(`SELECT ${targetColumns.join(", ")} FROM subscriptions
    WHERE user_id = ? AND (${predicates.join(" OR ")}) ORDER BY created_at DESC, id DESC LIMIT ?`)
    .bind(...values, SUBSCRIPTION_INDEX_LIMIT + 1).all<VisibilityTarget>();
  if (result.results.length > SUBSCRIPTION_INDEX_LIMIT) {
    throw new HttpError(422, serverText(locale, "common.invalidRequestParameters"), "SUBSCRIPTION_COLLECTION_LIMIT_EXCEEDED", { limit: SUBSCRIPTION_INDEX_LIMIT });
  }
  return result.results;
}

function visibilitySnapshotQuery(env: Env, userId: string, rows: readonly VisibilityTarget[]) {
  const matches = snapshotColumns.map((column, i) => `s.${column} IS json_extract(expected.value, '$[${i}]')`).join(" AND ");
  // CROSS JOIN 固定 JSON 目标先行；owner 索引先行会让每条事实重新扫描整个 JSON 集合。
  const sql = `WITH expected AS (SELECT value FROM json_each(?))
    SELECT COUNT(*) AS fact_count, COALESCE(SUM(CASE WHEN i.user_id = s.user_id AND i.public_hidden = s.public_hidden THEN 1 ELSE 0 END), 0) AS index_count
    FROM expected CROSS JOIN subscriptions AS s ON s.id = json_extract(expected.value, '$[0]')
    LEFT JOIN subscription_list_index AS i ON i.subscription_id = s.id WHERE s.user_id = ? AND ${matches}`;
  return { sql, values: [JSON.stringify(rows.map((row) => snapshotColumns.map((column) => row[column]))), userId] };
}

export async function writePublicVisibility(env: Env, userId: string, rows: readonly VisibilityTarget[], publicHidden: boolean, locale: AppLocale): Promise<void> {
  if (rows.length === 0) return;
  if (rows.length > SUBSCRIPTION_INDEX_LIMIT || rows.some((row) => row.user_id !== userId)) throw new Error("Invalid visibility mutation owner or size");
  const snapshot = visibilitySnapshotQuery(env, userId, rows);
  const ids = JSON.stringify(rows.map((row) => row.id));
  const target = boolToInt(publicHidden);
  const timestamp = nowIso();
  // 整个请求只提交一个 batch；按 200 条分别提交会留下前批成功、后批失败的半完成状态。
  // JSON1 将 5000 条压成固定四条 SQL；前后 guard 都在事务内，不能在 await batch 之后才发现索引缺行。
  const statements = [
    env.DB.prepare(`SELECT CASE WHEN fact_count = ? AND index_count = ? THEN 1 ELSE json('visibility_snapshot_conflict') END FROM (${snapshot.sql})`)
      .bind(rows.length, rows.length, ...snapshot.values),
    env.DB.prepare(`UPDATE subscriptions SET public_hidden = ?, updated_at = ? WHERE user_id = ? AND id IN (SELECT value FROM json_each(?))`)
      .bind(target, timestamp, userId, ids),
    env.DB.prepare(`UPDATE subscription_list_index SET public_hidden = ?, updated_at = ? WHERE user_id = ? AND subscription_id IN (SELECT value FROM json_each(?))`)
      .bind(target, timestamp, userId, ids),
    env.DB.prepare(`SELECT CASE WHEN COUNT(*) = ? THEN 1 ELSE json('visibility_projection_conflict') END
      FROM json_each(?) AS selected CROSS JOIN subscriptions AS s ON s.id = selected.value
      JOIN subscription_list_index AS i ON i.subscription_id = s.id
      WHERE s.user_id = ? AND i.user_id = s.user_id AND s.public_hidden = ? AND i.public_hidden = s.public_hidden
        AND s.updated_at = ? AND i.updated_at = s.updated_at`).bind(rows.length, ids, userId, target, timestamp),
  ];
  try {
    await env.DB.batch(statements);
  } catch (error) {
    // SQLite 的 guard 异常没有稳定业务码；回滚后只读比较事实快照，不能依赖驱动报错字符串判定 409。
    const current = await env.DB.prepare(snapshot.sql).bind(...snapshot.values)
      .first<{ fact_count: number; index_count: number }>().catch(() => null);
    if (current && current.fact_count !== rows.length) throw new HttpError(409, serverText(locale, "subscription.visibilityConflict"), "SUBSCRIPTION_WRITE_CONFLICT");
    throw visibilityError(error, locale);
  }
}

function visibilityError(error: unknown, locale: AppLocale): HttpError {
  if (error instanceof HttpError) return error;
  const requestId = crypto.randomUUID();
  console.error("subscription_public_visibility_failed", { requestId, error });
  return new HttpError(500, serverText(locale, "subscription.visibilityFailed"), "SUBSCRIPTION_PUBLIC_VISIBILITY_FAILED", { requestId });
}
